compiler/graindoc/graindoc.re

```diff
@@ -222,13 +223,87 @@ let generate_docs =
   ();
 };
 
-let graindoc = opts => {
-  let program = compile_typed(opts);
-  try(generate_docs(opts, program)) {
-  | exn =>
-    Format.eprintf("@[%s@]@.", Printexc.to_string(exn));
-    exit(2);
+type run = {
+  input_path: Fp.t(Fp.absolute),
+  output_path: option(Fp.t(Fp.absolute)),
+};
+
+let enumerate_directory = (input_dir_path, output_dir_path) => {
+  let all_files = Array.to_list(Fs_access.readdir(input_dir_path));
+  let grain_files =
+    List.filter(
+      filepath => Filename.extension(Fp.toString(filepath)) == ".gr",
+      all_files,
+    );
+  List.map(
+    filepath => {
+      // We relativize between the input directory and the full filepath
+      // such that we can reconstruct the directory structure of the input directory
+      let relative_path =
+        Fp.relativizeExn(~source=input_dir_path, ~dest=filepath);
+      let gr_basename = Option.get(Fp.baseName(relative_path));
+      let md_basename =
+        Filepath.String.remove_extension(gr_basename) ++ ".md";
+      let dirname = Fp.dirName(relative_path);
+      let md_relative_path = Fp.join(dirname, Fp.relativeExn(md_basename));
+      let output_path = Fp.join(output_dir_path, md_relative_path);
+      {input_path: filepath, output_path: Some(output_path)};
+    },
+    grain_files,
+  );
+};
+
+let enumerate_runs = opts =>
+  switch (opts.input, opts.output) {
+  | (File(input_file_path), None) =>
+    `Ok([{input_path: input_file_path, output_path: None}])
+  | (File(input_file_path), Some(Exists(File(output_file_path)))) =>
+    `Ok([
+      {input_path: input_file_path, output_path: Some(output_file_path)},
+    ])
+  | (File(input_file_path), Some(NotExists(output_file_path))) =>
+    `Ok([
+      {input_path: input_file_path, output_path: Some(output_file_path)},
+    ])
+  | (Directory(_), None) =>
+    `Error((
+      false,
+      "Directory input must be used with `-o` flag to specify output directory",
+    ))
+  | (Directory(input_dir_path), Some(Exists(Directory(output_dir_path)))) =>
+    `Ok(enumerate_directory(input_dir_path, output_dir_path))
+  | (Directory(input_dir_path), Some(NotExists(output_dir_path))) =>
+    `Ok(enumerate_directory(input_dir_path, output_dir_path))
+  | (File(input_file_path), Some(Exists(Directory(output_dir_path)))) =>
+    `Error((
+      false,
+      "Using a file as input cannot be combined with directory output",
+    ))
```

Comment:
I'd just create the file in that directory in this case.

Reply:
I started implementing that before I submitted this and it has some **very** non-intuitive behavior.

Since the directory specified is handled recursively, we calculate the "base path" of the contents, such that:
`grain doc stdlib -o foo` will produce a tree containing:
```
$ tree -d foo
foo
├── runtime
│   ├── unsafe
│   └── utils
└── sys
```
This allows us to run the command with the input and output directory as the same path:
`grain doc stdlib -o stdlib` and it will output each markdown file next to the grain file.

Now, if we implement what you are suggesting, the behavior becomes non-intuitive because you can't calculate a base path:
`grain stdlib/runtime/unsafe/constants.gr -o stdlib` would need to create `stdlib/constants.md` because we don't have a base path. You could say that it should use the `-o` directory as the base path, and that would work if the input and output are in the same directory, but what if you wanted it to go somewhere else?
`grain stdlib/runtime/unsafe/constants.gr -o foo` can't use `foo` as a base path, otherwise it'll create a `stdlib` directory inside of it, which is different behavior than directory + directory.

I don't like any of these non-intuitive behaviors and decided to disallow it with this error.

Reply:
The only way I could see this being supported **and** intuitive is if we add a `--recursive` flag which can only be used with directory input. But I really don't want that extra flag.

Reply:
Yup, that sounds reasonable.